### DISCUSSION TITLE
Using PropertyGrid for displaying own data

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#2730] [Q&A] [answered]
I would like to use PropertyGrid for displaying data that I get from querying a database provided by our client.  Examples (Classifiers and Property Formatting) I found in frontend-showcase seem to use data from the imodel and depend on Presentation class and/or PresentationPropertyDataProvider class.  Is there an example of using PropertyGrid with generic data or data that is not from imodel?  If so, please provide link or guidance on how I might be able to use PropertyGrid for my own data.

Thanks,

Ovais

**Comment by grigasp** [comment, marked as answer]
The `PropertyGrid` component merely shows a list of property records grouped by categories. The [PropertyRecord](https://github.com/iTwin/itwinjs-core/blob/9a083a25cb1fb61e6cf88da767459ba6e33f7af6/ui/appui-abstract/src/appui-abstract/properties/Record.ts#L33) object contains information about the property (label, type, unit, etc.) and the value. To feed your data into the property grid you have to create an [IPropertyDataProvider](https://github.com/iTwin/itwinjs-core/blob/9a083a25cb1fb61e6cf88da767459ba6e33f7af6/ui/components-react/src/components-react/propertygrid/PropertyDataProvider.ts#L58) implementation that returns a [PropertyData](https://github.com/iTwin/itwinjs-core/blob/9a083a25cb1fb61e6cf88da767459ba6e33f7af6/ui/components-react/src/components-react/propertygrid/PropertyDataProvider.ts#L30) object with your categories and records. 

This may be helpful as an example: https://github.com/iTwin/itwinjs-core/blob/master/test-apps/ui-test-app/src/frontend/appui/widgets/PropertyGridDemoWidget.tsx

**Reply by ovaisa** [reply]
Thank you @grigasp, I am able to use this code to achieve my objective of showing generic data from other sources.  I would appreciate guidance on one more detail.

In my grid, the value of the property is shown below the property label (as shown below - from Property Formatter sample).  I would like to show the values in front of the property label (as shown in Classifiers sample - image below).  Can you please guide how I can achieve this. 

Thanks again for your guidance.

![grid-vertical](https://user-images.githubusercontent.com/83007711/142256132-a20408c2-bf59-4459-9ed6-f57ec4c46dac.JPG)

![grid-horizontal](https://user-images.githubusercontent.com/83007711/142256304-f3ac80b6-151f-4ad2-af9f-68ded7ab9518.JPG)

**Reply by bsteinbk** [reply]
@ovaisa The PropertyGrid takes the following prop `orientation?: Orientation`. I believe if  you set that prop to Orientation.Horizontal the label and value should appear on same line.

**Reply by ovaisa** [reply]
I was able to achieve this by using orientation and isOrientationFixed properties.  Thanks
